adapters/ninthdecimal/ninthdecimal.go

```diff
@@ -0,0 +1,250 @@
+package ninthdecimal
+
+import (
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"text/template"
+
+	"github.com/golang/glog"
+	"github.com/mxmCherry/openrtb"
+	"github.com/prebid/prebid-server/adapters"
+	"github.com/prebid/prebid-server/errortypes"
+	"github.com/prebid/prebid-server/macros"
+	"github.com/prebid/prebid-server/openrtb_ext"
+)
+
+type NinthdecimalAdapter struct {
+	EndpointTemplate template.Template
+}
+
+//MakeRequests prepares request information for prebid-server core
+func (adapter *NinthdecimalAdapter) MakeRequests(request *openrtb.BidRequest, reqInfo *adapters.ExtraRequestInfo) ([]*adapters.RequestData, []error) {
+	errs := make([]error, 0, len(request.Imp))
+	if len(request.Imp) == 0 {
+		errs = append(errs, &errortypes.BadInput{Message: "No impression in the bid request"})
+		return nil, errs
+	}
+	imps, impExts, err := getImpressionsInfo(request.Imp)
+	if len(imps) == 0 {
+		return nil, err
+	}
+	errs = append(errs, err...)
+
+	pub2impressions, dispErrors := dispatchImpressions(imps, impExts)
+	errs = append(errs, dispErrors...)
+	if len(pub2impressions) == 0 {
+		return nil, errs
+	}
+
+	result := make([]*adapters.RequestData, 0, len(pub2impressions))
+	for k, imps := range pub2impressions {
+		bidRequest, err := adapter.buildAdapterRequest(request, &k, imps)
+		if err != nil {
+			errs = append(errs, err)
+			return nil, errs
+		} else {
+			result = append(result, bidRequest)
+		}
+	}
+	return result, errs
+}
+
+// getImpressionsInfo checks each impression for validity and returns impressions copy with corresponding exts
+func getImpressionsInfo(imps []openrtb.Imp) ([]openrtb.Imp, []openrtb_ext.ExtImpNinthdecimal, []error) {
+	errors := make([]error, 0, len(imps))
+	resImps := make([]openrtb.Imp, 0, len(imps))
+	resImpExts := make([]openrtb_ext.ExtImpNinthdecimal, 0, len(imps))
+
+	for _, imp := range imps {
+		impExt, err := getImpressionExt(&imp)
+		if err != nil {
+			errors = append(errors, err)
+			continue
+		}
+		if err := validateImpression(&imp, impExt); err != nil {
+			errors = append(errors, err)
+			continue
+		}
+		resImps = append(resImps, imp)
+		resImpExts = append(resImpExts, *impExt)
+
+	}
+	return resImps, resImpExts, errors
+}
+
+func validateImpression(imp *openrtb.Imp, impExt *openrtb_ext.ExtImpNinthdecimal) error {
+	if impExt.PublisherID == "" {
+		return &errortypes.BadInput{Message: "No pubid value provided"}
+	}
+	return nil
+}
+
+//Group impressions by ninthdecimal-specific parameters `pubid
+func dispatchImpressions(imps []openrtb.Imp, impsExt []openrtb_ext.ExtImpNinthdecimal) (map[openrtb_ext.ExtImpNinthdecimal][]openrtb.Imp, []error) {
+	res := make(map[openrtb_ext.ExtImpNinthdecimal][]openrtb.Imp)
+	errors := make([]error, 0)
+	for idx, imp := range imps {
+		err := compatImpression(&imp)
```

Comment:
Both `getImpressionsInfo` and this function loop through the `[]Imp` array can we do it in one loop? Maybe in a single function? I realize this doesn't decrease the time complexity, but it might be a small, feasible optimization.

Reply:
done